### PR TITLE
Add `--zap` flag for `brew uninstall`.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

Implies `brew uninstall` for formulae and `brew cask zap` for casks.

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
I think I'd rather this be `uninstall --zap` instead. Thoughts?

**Comment by reitermarkus** [comment]
Yeah, `brew uninstall --zap` sound good. 

cc @Homebrew/cask

File: Library/Homebrew/cmd/uninstall.rb [hunk 1]
```diff
@@ -19,12 +19,16 @@ module Homebrew
...
 
-        Uninstall <formula>.
+        Uninstall a <formula> or <cask>.
       EOS
       switch "-f", "--force",
-             description: "Delete all installed versions of <formula>."
+             description: "Delete all installed versions of <formula>. Uninstall even if <cask> is not " \
+                          "installed, overwrite existing files and ignore errors when removing files."
+      switch "--zap",
+             description: "Remove all files associated with the given <formula> or <cask>. " \
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/uninstall.rb:30]:
```suggestion
             description: "Remove all files associated with the given <cask>. " \
```

and conflict with `--formula` until this is supported?

File: docs/Manpage.md [hunk 2]
```diff
@@ -677,6 +677,22 @@ specify *`formula`* as a required or recommended dependency for their stable bui
 * `--skip-recommended`:
   Skip all formulae that specify *`formula`* as `:recommended` type dependency.
 
+### `zap` [*`options`*] *`formula`*|*`cask`*
```
Comment (MikeMcQuaid) [inline docs/Manpage.md:680]: Not this line but: should `brew cask zap` perhaps be removed from the manpage if it's deprecated (same with other commands)? That can be another PR, though.

File: Library/Homebrew/cmd/uninstall.rb (hunk 1 above)
Comment (reitermarkus) [reply Library/Homebrew/cmd/uninstall.rb:30]: Well, `--zap` for casks also implies `uninstall`, so I think the same should be the case for formulae, even if no other files are removed yet for formulae.

**Comment by vitorgalvao** [comment]
> Yeah, `brew uninstall --zap` sounds good.

Agreed. One thing to keep in mind is to make sure it continues to work with `--force` even if the cask was already removed.

File: docs/Manpage.md (hunk 2 above)
Comment (reitermarkus) [reply docs/Manpage.md:680]: I think it will be removed when adding it do the `DEPRECATED_COMMANDS` list. I think we should deprecate the remaining commands in bulk in another PR.

File: Library/Homebrew/cmd/uninstall.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/uninstall.rb:30]: I think people will get confused if it does nothing for formulae, though.

File: docs/Manpage.md (hunk 2 above)
Comment (MikeMcQuaid) [reply docs/Manpage.md:680]: 👍🏻 Cool. If it doesn't: it'd be good to add that as part of that PR.

File: Library/Homebrew/cmd/uninstall.rb (hunk 1 above)
Comment (vitorgalvao) [reply Library/Homebrew/cmd/uninstall.rb:30]: Couldn’t we just add a `zap isn’t yet supported on formulae` message when applicable?

Comment (reitermarkus) [reply Library/Homebrew/cmd/uninstall.rb:30]: Maybe a warning that it's a no-op for now.

Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/uninstall.rb:30]: I think it would be good to have a hard failure until it does anything. We have no plans to make `zap` work for formulae and it'll be a non-trivial amount of work to get there (and changing this flag back to reference formulae would be the easiest part of it).

File: Library/Homebrew/cmd/uninstall.rb [hunk 3]
```diff
@@ -19,12 +19,17 @@ module Homebrew
...
+        Uninstall a <formula> or <cask>.
       EOS
       switch "-f", "--force",
-             description: "Delete all installed versions of <formula>."
+             description: "Delete all installed versions of <formula>. Uninstall even if <cask> is not " \
+                          "installed, overwrite existing files and ignore errors when removing files."
+      switch "--zap",
+             description: "Remove all files associated with a <cask>. " \
+                          "*May remove files which are shared between applications.*"
+      conflicts "--formula", "--zap"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/uninstall.rb:32]: @reitermarkus Thanks for this 🙇🏻 ❤️
